timestep.c

```diff
@@ -132,218 +179,193 @@ void advance_and_find_timesteps(void)
+    double vv=0;
+    for(j=0; j < 3; j++)
+        vv += P[i].Vel[j] * P[i].Vel[j];
+    vv = sqrt(vv);
+    if(vv > All.MaxGasVel * velfac)
+        for(j=0;j < 3; j++)
+        {
+            P[i].Vel[j] *= All.MaxGasVel * velfac / vv;
             SPHP(i).VelPred[j] =
                 P[i].Vel[j] - dt_gravkick2 * P[i].GravAccel[j] - dt_hydrokick2 * SPHP(i).HydroAccel[j];
 
             SPHP(i).VelPred[j] += P[i].GravPM[j] * dt_gravkickB;
         }
 
-        const double velfac = sqrt(All.cf.a3inv);
-        double vv=0;
-        for(j=0; j < 3; j++)
-            vv += P[i].Vel[j] * P[i].Vel[j];
-        vv = sqrt(vv);
-        if(vv > All.MaxGasVel * velfac)
-            for(j=0;j < 3; j++)
-            {
-                P[i].Vel[j] *= All.MaxGasVel * velfac / vv;
-                SPHP(i).VelPred[j] =
-                    P[i].Vel[j] - dt_gravkick2 * P[i].GravAccel[j] - dt_hydrokick2 * SPHP(i).HydroAccel[j];
-
-                SPHP(i).VelPred[j] += P[i].GravPM[j] * dt_gravkickB;
-            }
+    /* In case of cooling, we prevent that the entropy (and
+       hence temperature) decreases by more than a factor 0.5.
+       FIXME: Why is this and the last thing here? Should not be needed. */
 
-        /* In case of cooling, we prevent that the entropy (and
-           hence temperature decreases by more than a factor 0.5 */
+    if(SPHP(i).DtEntropy * dt_entr > -0.5 * SPHP(i).Entropy)
+        SPHP(i).Entropy += SPHP(i).DtEntropy * dt_entr;
+    else
+        SPHP(i).Entropy *= 0.5;
 
-        if(SPHP(i).DtEntropy * dt_entr > -0.5 * SPHP(i).Entropy)
-            SPHP(i).Entropy += SPHP(i).DtEntropy * dt_entr;
-        else
-            SPHP(i).Entropy *= 0.5;
-
-        if(All.MinEgySpec)
+    /* Implement an entropy floor*/
+    if(All.MinEgySpec)
+    {
+        const double minentropy = All.MinEgySpec * GAMMA_MINUS1 / pow(SPHP(i).EOMDensity * All.cf.a3inv, GAMMA_MINUS1);
+        if(SPHP(i).Entropy < minentropy)
         {
-            minentropy = All.MinEgySpec * GAMMA_MINUS1 / pow(SPHP(i).EOMDensity * All.cf.a3inv, GAMMA_MINUS1);
-            if(SPHP(i).Entropy < minentropy)
-            {
-                SPHP(i).Entropy = minentropy;
-                SPHP(i).DtEntropy = 0;
-            }
+            SPHP(i).Entropy = minentropy;
+            SPHP(i).DtEntropy = 0;
         }
+    }
 
-        /* In case the timestep increases in the new step, we
-           make sure that we do not 'overcool'. */
-        dt_entr = (P[i].TimeBin ? (1 << P[i].TimeBin) : 0) / 2 * All.Timebase_interval;
+    /* In case the timestep increases in the new step, we
+       make sure that we do not 'overcool'. */
+    dt_entr = (P[i].TimeBin ? (1 << P[i].TimeBin) : 0) / 2 * All.Timebase_interval;
 
-        if(SPHP(i).Entropy + SPHP(i).DtEntropy * dt_entr < 0.5 * SPHP(i).Entropy)
-            SPHP(i).DtEntropy = -0.5 * SPHP(i).Entropy / dt_entr;
-    }
+    if(SPHP(i).Entropy + SPHP(i).DtEntropy * dt_entr < 0.5 * SPHP(i).Entropy)
```

Comment:
It does - the duplicate is there because the timestep and entropy floor may change the condition. It seems it would be best removed. But I do not understand why the condition is there at all, so I was not sure whether to touch it. Do you understand?